packages/scope-manager/src/referencer/TypeVisitor.ts

```diff
@@ -199,7 +199,21 @@ class TypeVisitor extends Visitor {
   protected TSMappedType(node: TSESTree.TSMappedType): void {
     // mapped types key can only be referenced within their return value
     this.#referencer.scopeManager.nestMappedTypeScope(node);
-    this.visitChildren(node);
+    this.#referencer
+      .currentScope()
+      .defineIdentifier(
+        node.key,
+        new TypeDefinition(
+          node.key,
+          node.parent as
+            | TSESTree.TSInterfaceDeclaration
+            | TSESTree.TSTypeAliasDeclaration
+            | TSESTree.TSTypeParameter,
```

Comment:
RC: the `TypeDefinition` class should be expanded to include `TSMappedType` here:
https://github.com/typescript-eslint/typescript-eslint/blob/210e2810f6e83be0f3b565cb16bf476a05d945dd/packages/scope-manager/src/definition/TypeDefinition.ts#L8-L10

```suggestion
          node,
```